config/initializers/rack_attack.rb

```diff
@@ -81,25 +85,33 @@ def self.protected_route?(protected_actions, path, method)
   # throttle logins for another user and force their login requests to be
   # denied, but that's not very common and shouldn't happen to you. (Knock
   # on wood!)
-  throttle("logins/handler", limit: REQUEST_LIMIT, period: LIMIT_PERIOD) do |req|
-    if req.path == "/session" && req.post?
-      # return the handler if present, nil otherwise
-      req.params['session']['who'].presence if req.params['session']
+  protected_sessions_action = [{ controller: "sessions", action: "create" }]
+
+  throttle("logins/handle", limit: REQUEST_LIMIT, period: LIMIT_PERIOD) do |req|
+    req.params['session']['who'].presence if protected_route?(protected_sessions_action, req.path, req.request_method) && req.params['session']
+  end
+
+  throttle("api_key/basic_auth", limit: REQUEST_LIMIT, period: LIMIT_PERIOD) do |req|
+    if protected_route?(protected_api_key_action, req.path, req.request_method)
+      action_dispatch_req = ActionDispatch::Request.new(req.env)
+      ActionController::HttpAuthentication::Basic.user_name_and_password(action_dispatch_req).first.presence
     end
   end
 
   ############################# rate limit per email ############################
+  protected_passwords_action = [{ controller: "passwords", action: "create" }]
+
   throttle("password/email", limit: REQUEST_LIMIT_PER_EMAIL, period: LIMIT_PERIOD) do |req|
-    if req.path == "/passwords" && req.post?
-      # return the email if present, nil otherwise
-      req.params['password']['email'].presence if req.params['password']
+    if protected_route?(protected_passwords_action, req.path, req.request_method) && req.params['password']
+      User.normalize_email(req.params['password']['email'].presence)
```

Comment:
I suspect the `.presence` call should be the last method call in the chain, so blank strings are treated as nil and do not persist a throttle in memcached.

```diff
-User.normalize_email(req.params['password']['email'].presence)
+User.normalize_email(req.params['password']['email']).presence
```